### ISSUE TITLE
update-python-libraries should use PyPi for determining latest versions

### ISSUE THREAD

**Issue by rhendric** [NixOS/nixpkgs#230424]
### Describe the bug

Almost half of Nixpkgs's Python library packages use `fetchFromGitHub` as their source. When `update-python-libraries` handles such packages, it checks the list of formal releases on GitHub for the latest version number. However, it's not uncommon for Python developers to treat making a GitHub release as an afterthought; publishing to PyPi is the important thing. This results in Nixpkgs missing out on automated version updates that we could be getting if we checked PyPi in all cases.

### Steps To Reproduce
A representative example (from recent r-ryantm logs, not one I specifically need updated):
1. Compare https://github.com/marshallward/f90nml/releases to https://pypi.org/project/f90nml/#history
2. Observe that https://github.com/marshallward/f90nml/releases/tag/v1.4.3 is available for us to fetch from if we had the actual latest version number from PyPi

### Expected behavior

I propose that `update-python-libraries` always look to PyPi to find the latest version number of a Python library. When a derivation uses a fetcher other than `fetchPypi`, it would have to infer the correct name to use—I think we should check for a field like `meta.pypiName` (naming suggestions welcome) on the derivation and use `pname` as a default if one isn't specified. Optionally GitHub-based derivations could fall back to checking releases there if the PyPi check fails.

### Notify maintainers

@FRidh @mweinelt @jonringer

**Comment by mweinelt** [comment]
The specific example isn't really great. I'm not sure why it fails like this, but it does detect the 1.4.3 tag.

https://r.ryantm.com/log/python3.10-f90nml/2023-02-15.log
https://r.ryantm.com/log/python3.10-f90nml/2023-03-06.log

**Comment by rhendric** [comment]
Yeah, to be clear, I'm volunteering to do the work; I just wanted feedback on the idea first.

**Comment by FRidh** [comment]
I agree somewhat. There are reasons why we also fetch from GitHub, and that often is because the sdist misses tests. In those cases having an issue opened might be more fruitful. At the same time, when going there one might as well use repology?

**Comment by rhendric** [comment]
I don't think we should stop actually fetching from GitHub. For GitHub-fetched libraries, I think we should get the list of tags from GitHub instead of the list of releases, get the latest version from PyPI (because there might be tags that don't correspond to releases), find the tag that, after stripping prefixes, matches the version we got from PyPI, and fetch that tag from GitHub per usual.

I guess we could use Repology for this instead of PyPI? But it'd be one more layer of indirection and potential delay, and I don't really see what it gets us.
